Add #[rustc_legacy_const_generics]

Amanieu opened the pull request:
This is the first step towards removing `#[rustc_args_required_const]`: a new attribute is added which rewrites function calls of the form `func(a, b, c)` to `func::<{b}>(a, c)`. This allows previously stabilized functions in `stdarch` which use `rustc_args_required_const` to use const generics instead.

This new attribute is not intended to ever be stabilized, it is only intended for use in `stdarch` as a replacement for `#[rustc_args_required_const]`.

```rust
#[rustc_legacy_const_generics(1)]
pub fn foo<const Y: usize>(x: usize, z: usize) -> [usize; 3] {
    [x, Y, z]
}

fn main() {
    assert_eq!(foo(0 + 0, 1 + 1, 2 + 2), [0, 2, 4]);
    assert_eq!(foo::<{1 + 1}>(0 + 0, 2 + 2), [0, 2, 4]);
}
```

r? @oli-obk

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1963404186) [(plain)](https://github.com/rust-lang/rust/commit/d87eec1bf67959977c8ebaadc2ddcf4bdc14658a/checks/1963404186/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
   Compiling regex v1.4.3
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 8.04s
tidy check
tidy error: following path contains more than 1418 entries, you should move the test to some relevant subdirectory (current: 1419): /checkout/src/test/ui
Found 436 error codes
Found 0 error codes with no tests
Done!
Done!
tidy error: /checkout/src/test/ui/legacy-const-generics.rs:16: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_ast_lowering/src/lib.rs:2834: TODO is deprecated; use FIXME
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

Aaron1011 commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
+                    arg.span,
+                );
+
+                let anon_const = AnonConst { id: node_id, value: arg };
```
Could you add a test for the case where the argument is not a valid constant (e.g. `let a = 1; foo(0, a, 0)`)?

Amanieu replied:
The compiler ICEs in this case: late resolution happens before AST lowering and needs to be aware of the rewrite since the expression in the `AnonConst` needs to be resolved inside a `ConstantItemRibKind`.

I will fix this and add a test case.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1963496223) [(plain)](https://github.com/rust-lang/rust/commit/e604d01fede6a272f5cb4aae283a7db2219f0adc/checks/1963496223/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 436 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/test/ui/legacy-const-generics.rs:16: TODO is deprecated; use FIXME
tidy error: /checkout/compiler/rustc_ast_lowering/src/lib.rs:2834: TODO is deprecated; use FIXME
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

Amanieu commented:
The current implementation is not very well optimized, let's see how bad it is...

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 00eca69bffcb4ef95ba1dd6013b840bf12a85804 with merge 6d7225329e02f2f6fc73dfb977e802102a412bff...
<!-- homu: {"type":"TryBuildStarted","head_sha":"00eca69bffcb4ef95ba1dd6013b840bf12a85804","merge_sha":"6d7225329e02f2f6fc73dfb977e802102a412bff"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1968170957)
Build commit: 6d7225329e02f2f6fc73dfb977e802102a412bff (`6d7225329e02f2f6fc73dfb977e802102a412bff`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1968170957"},"merge_sha":"6d7225329e02f2f6fc73dfb977e802102a412bff"} -->

rust-timer commented:
Queued 6d7225329e02f2f6fc73dfb977e802102a412bff with parent fe1bf8e05c39bdcc73fc09e246b7209444e389bc, future [comparison URL](https://perf.rust-lang.org/compare.html?start=fe1bf8e05c39bdcc73fc09e246b7209444e389bc&end=6d7225329e02f2f6fc73dfb977e802102a412bff).

oli-obk commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
+    /// from the attribute.
+    fn legacy_const_generic_args(&mut self, expr: &Expr) -> Option<Vec<usize>> {
+        if let ExprKind::Path(None, path) = &expr.kind {
+            if path.segments.last().unwrap().args.is_some() {
```
Please leave a comment stating that this rewrite doesn't work for function calls with generic parameters specified.

Isn't this a problem as some of these functions already have type generics and there may be stable code out there in the wild that has calls with generic types specified?

Amanieu replied:
None of the functions using `rustc_args_required_const` are generic, they are all based on a vendor's C API.

Excluding generics is intentional: if you're trying to use generics then you are most likely trying to call the "new" form of the intrinsic and the rewrite shouldn't apply. I'll add a comment to clarify.

oli-obk commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
+            if path.segments.last().unwrap().args.is_some() {
+                return None;
+            }
+            if let Some(partial_res) = self.resolver.get_partial_res(expr.id) {
```
```suggestion
            let partial_res = self.resolver.get_partial_res(expr.id)?;
```

oli-obk commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
+        let mut generic_args = vec![];
+        for (idx, arg) in args.into_iter().enumerate() {
+            if legacy_args_idx.contains(&idx) {
+                let parent_def_id = self.current_hir_id_owner.last().unwrap().0;
+                let node_id = self.resolver.next_node_id();
+
+                // Add a definition for the in-band const def.
+                self.resolver.create_def(
+                    parent_def_id,
+                    node_id,
+                    DefPathData::AnonConst,
+                    ExpnId::root(),
+                    arg.span,
+                );
+
+                let anon_const = AnonConst { id: node_id, value: arg };
```
can we share some code with the regular anon const lowering here?

Amanieu replied:
Most of the code is already shared. The `AnonConst` isn't actually lowered here, it is just appended to the generic args of the path expression `f`. `f` is then lowered normally through `lower_expr`.

oli-obk commented on compiler/rustc_ast_lowering/src/lib.rs:
```diff
@@ -2826,6 +2828,16 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
+
+    fn item_attrs(&self, def_id: DefId) -> Vec<ast::Attribute> {
+        if let Some(_local_def_id) = def_id.as_local() {
+            // FIXME: This doesn't actually work, items doesn't include everything?
+            //self.items[&hir::ItemId { def_id: local_def_id }].attrs.into()
+            Vec::new()
```
Yea I think this is a non-issue. Any function that calls a rustc_legacy_const_generics function within the same crate can just use const generics as users will never see that attribute.

oli-obk commented on compiler/rustc_resolve/src/late.rs:
```diff
@@ -2406,6 +2418,42 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
+    /// Checks if an expression refers to a function marked with
+    /// `#[rustc_legacy_const_generics]` and returns the argument index list
+    /// from the attribute.
+    fn legacy_const_generic_args(&mut self, expr: &Expr) -> Option<Vec<usize>> {
```
This is an exact duplicate of the same function in lowering, right? If so, could it be implemented on Resolver, which I think both here and lowering have access to?

oli-obk commented on src/test/ui/auxiliary/legacy-const-generics.rs:
```diff
@@ -0,0 +1,6 @@
+#![feature(rustc_attrs)]
+
+#[rustc_legacy_const_generics(1)]
+pub fn foo<const Y: usize>(x: usize, z: usize) -> [usize; 3] {
+    [x, Y, z]
+}
```
please also add a test for a function which also has a generic type parameter for one of the runtime arguments

oli-obk reviewed (changes requested):
I think we should ignore the same-crate situation and just not do anything there. If you can write functions with const generics you can change your call sites, too.

rust-timer commented:
Finished benchmarking try commit (6d7225329e02f2f6fc73dfb977e802102a412bff): [comparison url](https://perf.rust-lang.org/compare.html?start=fe1bf8e05c39bdcc73fc09e246b7209444e389bc&end=6d7225329e02f2f6fc73dfb977e802102a412bff).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

oli-obk commented:
Ok, it does have a measurable impact. So... one way to optimize this would be to write the list of all `rustc_legacy_const_generics` of a crate into metadata. Then, instead of looking up the attribute, we can look into that list (should probably become a hash set upon reading from metadata). Since most crates' list will be entirely empty, this should be really fast to check.

Though... I'm not sure if this is actually a bottleneck at all since there seem to be no additional calls/decodings of `item_attrs`, though that may just be because these aren't registered yet at these early stages where we don't have tcx yet.

Amanieu commented:
I think part of the overhead comes from returning a `Vec` which requires an allocation on every call to `item_attrs`. I'll try changing that function to take a `FnMut(&[Attributes])` closure argument instead, which should avoid the allocation.

oli-obk commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -292,6 +298,86 @@ impl<'hir> LoweringContext<'_, 'hir> {
+    /// from the attribute.
+    fn legacy_const_generic_args(&mut self, expr: &Expr) -> Option<Vec<usize>> {
+        if let ExprKind::Path(None, path) = &expr.kind {
+            if path.segments.last().unwrap().args.is_some() {
+                return None;
+            }
```
since we don't have any such functions right now, the legacy support doesn't need to support it. If anyone creates a function like this, they'll just have to use const generics syntax. This PR is just so the call sites to existing functions keep working

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1974895942) [(plain)](https://github.com/rust-lang/rust/commit/2451f124c95b2edfcaac9706500163c0224502df/checks/1974895942/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 4100/11494
.................................................................................................... 4200/11494
.................................................................................................... 4300/11494
...................................................................ii............................... 4400/11494
....................................................i..F............................................ 4500/11494
.................................................................................................... 4700/11494
.................................................................................................... 4800/11494
.................................................................................................... 4900/11494
.................................................................................................... 5000/11494
---
.................................................................................................... 9200/11494
.................................................................................................... 9300/11494
.................................................................................................... 9400/11494
..................................................i......i.......................................... 9500/11494
.........................................................................................iiiiiii..ii 9600/11494
.................................................................................................... 9800/11494
.................................................................................................... 9900/11494
.................................................................................................... 10000/11494
.................................................................................................... 10100/11494
---
.....................................................................................i.i............ 11400/11494
..............................................................................................
failures:

---- [ui] ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs stdout ----


+ error[E0428]: the name `foo3` is defined multiple times
+   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:30:1
+    |
+ LL | fn foo3<const X: usize>(_: u8) {}
+    | ------------------------------ previous definition of the value `foo3` here
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+ LL | fn foo3<X>() {}
+    | ^^^^^^^^^^^^ `foo3` redefined here
+    |
+    = note: `foo3` must be defined only once in the value namespace of this module
+ 
1 error: suffixed literals are not allowed in attributes
2   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:21:31


7    = help: instead of using a suffixed literal (`1u8`, `1.0f32`, etc.), use an unsuffixed version (`1`, `1.0`, etc.)
8 
9 error: malformed `rustc_legacy_const_generics` attribute input
-   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:29:1
+   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:32:1
11    |
12 LL | #[rustc_legacy_const_generics]
13    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_legacy_const_generics(N)]`
14 
14 
15 error: malformed `rustc_legacy_const_generics` attribute input
-   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:32:1
+   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:35:1
17    |
18 LL | #[rustc_legacy_const_generics = 1]
19    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_legacy_const_generics(N)]`
56 LL | struct S;
57    | --------- not a function
58 
58 
+ error: #[rustc_legacy_const_generics] functions must only have const generics
+   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:29:31
+    |
+ LL | #[rustc_legacy_const_generics(0)]
+    |                               ^
+ LL | fn foo3<X>() {}
+    |         - non-const generic parameter
+ 
59 error: index exceeds number of arguments
60   --> $DIR/invalid-rustc_legacy_const_generics-arguments.rs:25:35


62 LL |     #[rustc_legacy_const_generics(1)]
63    |                                   ^ there is only 1 argument
- error: aborting due to 10 previous errors
+ error: aborting due to 12 previous errors
66 
+ For more information about this error, try `rustc --explain E0428`.
+ For more information about this error, try `rustc --explain E0428`.
67 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments/invalid-rustc_legacy_const_generics-arguments.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args invalid/invalid-rustc_legacy_const_generics-arguments.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0428]: the name `foo3` is defined multiple times
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:30:1
   |
LL | fn foo3<const X: usize>(_: u8) {}
   | ------------------------------ previous definition of the value `foo3` here
...
LL | fn foo3<X>() {}
   | ^^^^^^^^^^^^ `foo3` redefined here
   |
   = note: `foo3` must be defined only once in the value namespace of this module

error: suffixed literals are not allowed in attributes
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:21:31
   |
LL | #[rustc_legacy_const_generics(0usize)] //~ ERROR suffixed literals are not allowed in attributes
   |
   |
   = help: instead of using a suffixed literal (`1u8`, `1.0f32`, etc.), use an unsuffixed version (`1`, `1.0`, etc.)

error: malformed `rustc_legacy_const_generics` attribute input
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:32:1
   |
LL | #[rustc_legacy_const_generics] //~ ERROR malformed `rustc_legacy_const_generics` attribute
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_legacy_const_generics(N)]`

error: malformed `rustc_legacy_const_generics` attribute input
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:35:1
   |
LL | #[rustc_legacy_const_generics = 1] //~ ERROR malformed `rustc_legacy_const_generics` attribute
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_legacy_const_generics(N)]`

error: index exceeds number of arguments
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:3:31
   |
LL | #[rustc_legacy_const_generics(0)] //~ ERROR index exceeds number of arguments
   |                               ^ there are only 0 arguments

error: index exceeds number of arguments
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:6:31
   |
LL | #[rustc_legacy_const_generics(1)] //~ ERROR index exceeds number of arguments
   |                               ^ there is only 1 argument

error: index exceeds number of arguments
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:9:31
   |
LL | #[rustc_legacy_const_generics(2)] //~ ERROR index exceeds number of arguments
   |                               ^ there are only 2 arguments
error: arguments should be non-negative integers
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:12:31
   |
   |
LL | #[rustc_legacy_const_generics(a)] //~ ERROR arguments should be non-negative integers

error: arguments should be non-negative integers
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:15:34
   |
   |
LL | #[rustc_legacy_const_generics(1, a, 2, b)] //~ ERROR arguments should be non-negative integers
   |                                  ^     ^
error: attribute should be applied to a function
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:18:1
   |
   |
LL | #[rustc_legacy_const_generics(0)] //~ ERROR attribute should be applied to a function
LL | struct S;
   | --------- not a function


error: #[rustc_legacy_const_generics] functions must only have const generics
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:29:31
   |
LL | #[rustc_legacy_const_generics(0)] //~ ERROR #[rustc_legacy_const_generics] functions must only have
   |                               ^
LL | fn foo3<X>() {}
   |         - non-const generic parameter

error: index exceeds number of arguments
  --> /checkout/src/test/ui/invalid/invalid-rustc_legacy_const_generics-arguments.rs:25:35
   |
LL |     #[rustc_legacy_const_generics(1)] //~ ERROR index exceeds number of arguments
   |                                   ^ there is only 1 argument
error: aborting due to 12 previous errors

For more information about this error, try `rustc --explain E0428`.

---
test result: FAILED. 11400 passed; 1 failed; 93 ignored; 0 measured; 0 filtered out; finished in 133.06s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:13:50
```

</details>

Amanieu commented:
I've addressed the review comments and have attempted to address the performance issues using my favored strategy of throwing a `HashMap` at the problem.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 22184a0f5d0e83d07425d6058e7ad93c78160233 with merge 96bc946ebd85c25f0f05e1b979e8f996be1a1daf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"22184a0f5d0e83d07425d6058e7ad93c78160233","merge_sha":"96bc946ebd85c25f0f05e1b979e8f996be1a1daf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1975332565)
Build commit: 96bc946ebd85c25f0f05e1b979e8f996be1a1daf (`96bc946ebd85c25f0f05e1b979e8f996be1a1daf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1975332565"},"merge_sha":"96bc946ebd85c25f0f05e1b979e8f996be1a1daf"} -->

rust-timer commented:
Queued 96bc946ebd85c25f0f05e1b979e8f996be1a1daf with parent 1fdadbf13aecd190b277ea2aa1b125d2ed986d55, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1fdadbf13aecd190b277ea2aa1b125d2ed986d55&end=96bc946ebd85c25f0f05e1b979e8f996be1a1daf).

rust-timer commented:
Finished benchmarking try commit (96bc946ebd85c25f0f05e1b979e8f996be1a1daf): [comparison url](https://perf.rust-lang.org/compare.html?start=1fdadbf13aecd190b277ea2aa1b125d2ed986d55&end=96bc946ebd85c25f0f05e1b979e8f996be1a1daf).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

jyn514 commented:
Perf looks much better now :)

Amanieu commented:
@oli-obk Ready for review.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 00afbe70f2f5cb0b6455036f4b97537ff77f2052 has been approved by `oli-obk`

<!-- @bors r=oli-obk 00afbe70f2f5cb0b6455036f4b97537ff77f2052 -->
<!-- homu: {"type":"Approved","sha":"00afbe70f2f5cb0b6455036f4b97537ff77f2052","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 00afbe70f2f5cb0b6455036f4b97537ff77f2052 with merge 80c5a3d003b0f4610bddd0dd4237980f6f21d9a5...
<!-- homu: {"type":"BuildStarted","head_sha":"00afbe70f2f5cb0b6455036f4b97537ff77f2052","merge_sha":"80c5a3d003b0f4610bddd0dd4237980f6f21d9a5"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1978986020)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1978986020","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1978986020) [(plain)](https://github.com/rust-lang-ci/rust/commit/80c5a3d003b0f4610bddd0dd4237980f6f21d9a5/checks/1978986020/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

oli-obk commented:
A lot of builders passed, let's try again

@bors retry

bors commented:
:hourglass: Testing commit 00afbe70f2f5cb0b6455036f4b97537ff77f2052 with merge 98f8cce6db6c6c6660eeffee2b3903104e547ecf...
<!-- homu: {"type":"BuildStarted","head_sha":"00afbe70f2f5cb0b6455036f4b97537ff77f2052","merge_sha":"98f8cce6db6c6c6660eeffee2b3903104e547ecf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1982609221)
Approved by: oli-obk
Pushing 98f8cce6db6c6c6660eeffee2b3903104e547ecf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1982609221"},"merge_sha":"98f8cce6db6c6c6660eeffee2b3903104e547ecf"} -->
